examples/source_separation/utils/dist_utils.py

```diff
@@ -0,0 +1,74 @@
+import os
+import types
+import logging
+
+import torch
+import torch.distributed as dist
+
+
+def _info_on_master(self, *args, **kwargs):
+    if dist.get_rank() == 0:
+        self.info(*args, **kwargs)
+
+
+def getLogger(name):
+    """Get logging.Logger module with additional ``info_on_master`` method."""
+    logger = logging.getLogger(name)
+    logger.info_on_master = types.MethodType(_info_on_master, logger)
+    return logger
+
+
+_LG = getLogger(__name__)
+
+
+def setup_distributed(
+    world_size, rank, local_rank, backend="nccl", init_method="env://"
+):
+    """Perform env setup and initialization for distributed training"""
+    if init_method == "env://":
+        _set_env_vars(world_size, rank, local_rank)
+    params = {
+        "backend": backend,
+        "init_method": init_method,
+        "world_size": world_size,
+        "rank": rank,
+    }
+    _LG.info("Initializing distributed process group with %s", params)
+    dist.init_process_group(**params)
+    _LG.info("Initialized distributed process group.")
+
+
+def _set_env_vars(world_size, rank, local_rank):
+    for key, default in [("MASTER_ADDR", "127.0.0.1"), ("MASTER_PORT", "29500")]:
+        if key not in os.environ:
+            os.environ[key] = default
+
+    os.environ["WORLD_SIZE"] = str(world_size)
+    os.environ["RANK"] = str(rank)
+    os.environ["LOCAL_RANK"] = str(local_rank)
+
+
+def save_on_master(obj, path):
+    if dist.get_rank() == 0:
+        _LG.info("Saving %s", path)
+        torch.save(obj, path)
+
+
+def synchronize_params(path, *modules):
```

Comment:
If you don't run this function very often it's fine to do it as is, but it is possible to avoid serializing the model to a file and instead use distributed communications to perform the broadcasting for you. Probably something involving using pickle to serialize the dict.

Here is an example of what we did for a slightly different function (all_gather) https://github.com/pytorch/vision/blob/497744b9d510ff2df756f479ee5a19fce0d579b6/references/detection/utils.py#L75-L115

Reply:
This function is dedicated for the parameter initialization on training from scratch and based on the example. 

https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#save-and-load-checkpoints

I think the `gather_all` utility is a bit too much for this purpose.